### ISSUE TITLE
JavaScript timing in MiniProfiler results

### ISSUE THREAD

**Issue by StuffOfInterest** [MiniProfiler/dotnet#302]
I've seen scattered references but nothing concrete regarding capturing timing information in JavaScript code and having it displayed in the MiniProfiler results.  I'm dealing with issues where the DOM load is taking longer than the service side execute and render for pages.  The page uses several script libraries so it would be nice to capture timing information on script execution with something like:

```
window.MiniProfiler.StepIn("step name");
SomeLongRunningCode();
window.MiniProfiler.StepOut();
```

or maybe:

```
window.MiniProfiler.Step("step name", function() {
   SomeLongRunningCode();
});
```

Being that there are cases where loading the script needs to be timed, the former syntax may be more versatile.  That version would also be easier to suppress if MiniProfiler is not active (perhaps with a tag helper in cshtml).

**Comment by NickCraver** [comment]
You can indeed do this today, you'll want to init the timing script up top, e.g.:
```cshtml
@this.InitClientTimings()
```
And your code would look like:
```js
if (window.mPt) mPt.start("step name");
SomeLongRunningCode();
if (window.mPt) mPt.end("step name");
```
Or, there are helpers, like this:
```cshtml
@this.TimeScript("step name", @<script>SomeLongRunningCode();</script>)
```

You can also search the sample projects for `ClientTimingHelper` to see some examples of it in play.

**Comment by StuffOfInterest** [comment]
One small problem with this.  It lives in the MiniProfiler.Mvc5 project.  I'm working on an ASP.NET Core app.  That explains why I wasn't able to make the `InitClientTimings()` extension method work before.

Being that the core code for the helper is in a shared project (MiniProfiler.Shared) and the code for the extension methods is pretty straight forward, it shouldn't take too much to turn this into a couple of tag helpers for .NET Core.  I'll see what I can do there when I have some time available.  For now, I may just hack it into a project I'm working on to get a look at it.

**Comment by StuffOfInterest** [comment]
Aside from me being selectively blind this early in the morning and not seeing the new entry right away, it works. 

First, I put ```@Html.Raw(StackExchange.Profiling.ClientTimingHelper.InitScript)``` just below the ```<mini-profiler />``` at the top of the page body.  Next, I wrapped the part of the page which loads a bunch of other scripts with ```<script>mPt.start('Layout script load')</script>``` and ```<script>mPt.end('Layout script load')</script>```.  

The page loads and when I bring up the MiniProfiler output I see this:

![image](https://user-images.githubusercontent.com/8578647/41969674-3c00aff2-79d6-11e8-9708-8457b8eafd46.png)

Note the timing section for "Layout script load".

Definitely need to turn this into a couple of tag helpers now, but at least I can see into the script load process during my initial development.

**Comment by StuffOfInterest** [comment]
Created a pull request (#305) to adjust the colors for the client timing bar.  Hopefully this looks a little better, keeping in mind that I have no sense of style.

![image](https://user-images.githubusercontent.com/8578647/42104136-7acb1c94-7b99-11e8-968c-b6b701c51dab.png)

**Comment by NickCraver** [comment]
Finally getting back to this - resolved in 14025a4f155ff0e2c8eb146876c56071ad2cd3c7, the new color for unknown is `#80ccda`.

**Comment by StuffOfInterest** [comment]
Should I open another issue for the lack of indentation on nested client timing steps?

**Comment by NickCraver** [comment]
@StuffOfInterest It'd have to change the interface users have, so it'd be a breaking change either way (which means it'd not be until MiniProfiler v5), given the frequency of their usage I honestly don't picture it happening (it's also easy to break the tree). The serialization would be much more complex as well.

Remember that just because JavaScript is traditionally "single threaded" doesn't mean timings would be at all, you'd need something akin to `AsyncLocal<T>` behavior for tree tracking due to workers, fetch, etc. being things in browsers now...there's a lot of concurrent activity that can happen. Due to that, we can't realistically infer child relationships on the server-side just from start/finish timings either (e.g. "who's the parent?").

There are other problems with inference like the system clock changing during requests and such that also affects this, but IMO that's not worth diving into. It definitely happens though, we see it all day every day while recording timings like this form clients on Stack Overflow.

**Comment by StuffOfInterest** [comment]
I don't think were talking about the same thing here.  I wouldn't want to try and wrap my head around profiling parallel execution.  

The interface already allows nesting of client timing.  You can see that in my picture above where three of the bars (that have names starting with ">") are nested below the bar above.  That part is fine.  However, the names themselves are all presented at the same level.  In the code timing blocks (top area), if profiling sections are nested then the names are indented to show the nesting.

**Comment by StuffOfInterest** [comment]
Nested as in start-start-end-end.

![image](https://user-images.githubusercontent.com/8578647/43046637-d1b7d5da-8d99-11e8-8df2-e113bffe9ecd.png)

I have a large block of script wrapped with "start" and "end".  Inside of that block are three sub-blocks each wrapped with "start" and "end.  Nothing is happening in parallel here, I'm just trying to get roll-up numbers for some groups of scripts that are loading.  The only issue I have is that the text is not being indented in like happens when there are nested profiling blocks on the server side code.

**Comment by NickCraver** [comment]
@StuffOfInterest *for you* nothing is happening in parallel. That's not true for everyone, not even close. It breaks down immediately when that happens. If you try and implement this, you'll see what I'm talking about. *Which item you best under* necessitates tracking a parent. There's no way around this. The other thing you can do is try to infer based on timings and as I covered: that just doesn't work.

The items indented in the above server timings are *because it's a tree of timings*. It's not a *list* of timings, which is what JavaScript/client-side has.

**Comment by StuffOfInterest** [comment]
OK, last try here.  First, here is a capture of a set of timings I did using code based steps, profile tag helper to see rendering time, and the client timing tag helper for client execution.

![image](https://user-images.githubusercontent.com/8578647/43048601-220c9354-8db8-11e8-8e84-0f8e2b37c836.png)

Here is the code I used in my view.

```
<profile-script />
<profile name="Profile group">
	<profile name="Profile block 1">
		<profile-script name="JS delay outer">
			<profile-script name="JS inner 1">
				<script>
					alert("Wait for a second");
				</script>
			</profile-script>
			<profile-script name="JS inner 2">
				<script>
					alert("Wait for another second");
				</script>
			</profile-script>
		</profile-script>
	</profile>

	<profile name="Profile block 2">
		<div class="row">
			<div class="col-md-3">
				<h2>Application uses</h2>
				<ul>
					<li>Sample pages using ASP.NET Core MVC</li>
					<li>Theming using <a href="https://go.microsoft.com/fwlink/?LinkID=398939">Bootstrap</a></li>
				</ul>
			</div>
		</div>
	</profile>
</profile>
```

The only issue I'm talking about here is that the text "JS inner 1" and "JS inner 2" should be indented as their numbers are rolled into the "JS delay outer" total.  The same could probably be said about having "JS delay outer" indented as its total is rolled into the "DOM Loading" line immediately above.

Oh, and in the process of mocking this up I found a bug in the new profile-script tag helper.  The client timing initialization was being rendered with the second "profile-script" tag instead of the first.  Most likely this is due to the child content being rendered before the logic to add in the initialization code.  A little restructuring of the tag helper should fix it.  I'll try to take care of it soon.
